lib/puppet/network/resolver.rb

```diff
@@ -1,85 +1,134 @@
+      # current record plus previous records is greater than the random number.
+
+      total_weight = records.inject(0) { |sum,record|
+        sum + weight(record)
+      }
+      current_weight = 0
+      chosen_weight  = 1 + Kernel.rand(total_weight)
+
+      records.each do |record|
+        current_weight += weight(record)
+        return record if current_weight >= chosen_weight
+      end
     end
 
-    pri_hash.keys.sort.each do |key|
-      yield key, pri_hash[key]
+    def weight(record)
+      record.weight == 0 ? 1 : record.weight * 10
     end
-  end
-  private_class_method :each_priority
-
-  def self.find_weighted_server(records)
-    return nil if records.nil? || records.empty?
-    return records.first if records.size == 1
-
-    # Calculate the sum of all weights in the list of resource records,
-    # This is used to then select hosts until the weight exceeds what
-    # random number we selected.  For example, if we have weights of 1 8 and 3:
-    #
-    # |-|---|--------|
-    #        ^
-    # We generate a random number 5, and iterate through the records, adding
-    # the current record's weight to the accumulator until the weight of the
-    # current record plus previous records is greater than the random number.
-
-    total_weight = records.inject(0) { |sum,record|
-      sum + weight(record)
-    }
-    current_weight = 0
-    chosen_weight  = 1 + Kernel.rand(total_weight)
-
-    records.each do |record|
-      current_weight += weight(record)
-      return record if current_weight >= chosen_weight
+
+    # Returns TTL for the cached records for this service. If multiple
+    # records for this service were found, this returns the smallest
+    # TTL among them.
+    # @param [String] service_name the service whose TTL we want
+    # @return [Integer] the TTL for this service, in seconds
+    def ttl(service_name)
+      return @record_cache[service_name][:ttl]
     end
-  end
 
-  def self.weight(record)
-    record.weight == 0 ? 1 : record.weight * 10
+    # Checks if the cached entry for the given service has expired.
+    # @param [String] service_name the name of the service to check
+    # @return [Boolean] true if the entry has expired, false otherwise.
+    #                  Always returns true if the record had no TTL.
+    def expired?(service_name)
+      if ttl
+        return Time.now > (@record_cache[service_name][:resolved_at] + ttl)
+      else
+        return true
+      end
+    end
+
+    private
+
+    def update_cached_ttl(service_name, record)
+      cache_entry = @record_cache[service_name]
+      if cache_entry[:ttl].nil? || record.ttl < cache_entry[:ttl]
+        cache_entry[:ttl] = record.ttl
+      end
+    end
+
+    def each_priority(records)
+      pri_hash = records.inject({}) do |groups, element|
+        groups[element.priority] ||= []
+        groups[element.priority] << element
+        groups
+      end
+
+      pri_hash.keys.sort.each do |key|
+        yield key, pri_hash[key]
```

Comment:
We don't actually use the priority (key) do we?

Reply:
You're correct, it's not used by callers of `each_priority`. I think that's a remnant of a previous iteration of this code.